File: pc-frontend/src/features/project-docs/projectDocumentKnowledgeGraphModel.ts
export type ProjectKnowledgeMapView = 'capabilities' | 'architecture' | 'topics'
export type DocumentationStatus = 'documented' | 'partial' | 'undocumented'
export type ImplementationStatus = 'verified' | 'declared' | 'missing' | 'not_applicable'

export interface ProjectKnowledgeNodeConfig {
  id: string
  view: 'capabilities' | 'architecture'
  kind: string
  label: string
  detail: string
  parent_id: string
  order: number
  color: string
  entrypoint: string
  document_paths: string[]
  implementation_refs: string[]
  tags: string[]
}

export interface ProjectKnowledgeEdgeConfig {
  id: string
  source: string
  target: string
  relation: string
  label: string
}

export interface ProjectKnowledgeGraphConfig {
  nodes: ProjectKnowledgeNodeConfig[]
  edges: ProjectKnowledgeEdgeConfig[]
}

export interface ProjectKnowledgeMapEvidence {
  reference: string
  verification: 'exists' | 'missing' | 'declared'
}

export interface ProjectKnowledgeMapCoverage {
  key: 'overview' | 'requirements' | 'architecture' | 'reference' | 'operations' | 'evidence'
  label: string
  covered: boolean
  count: number
}

export interface ProjectKnowledgeMapNode {
  id: string
  view: ProjectKnowledgeMapView
  kind: string
  label: string
  detail: string
  color: string
  parent_id: string
  section_id: string
  depth: number
  child_count: number
  order: number
  document_count: number
  document_paths: string[]
  entrypoint: string
  entrypoint_source: 'configured' | 'inferred' | 'missing'
  coverage: ProjectKnowledgeMapCoverage[]
  missing_coverage: string[]
  documentation_status: DocumentationStatus
  implementation_refs: ProjectKnowledgeMapEvidence[]
  implementation_status: ImplementationStatus
  source: 'manifest' | 'profile_template'
  tags: string[]
}

export interface ProjectKnowledgeMapEdge {
  id: string
  source: string
  target: string
  relation: string
  label: string
  configured: boolean
}

export interface ProjectKnowledgeMapFinding {
  code: string
  severity: 'error' | 'warning' | 'info'
  node_id: string
  message: string
  suggested_action: string
}

export interface ProjectKnowledgeMap {
  version: 1
  view: ProjectKnowledgeMapView
  title: string
  source: 'manifest' | 'profile_template'
  root_id: string
  nodes: ProjectKnowledgeMapNode[]
  edges: ProjectKnowledgeMapEdge[]
  stats: {
    nodes: number
    configured_nodes: number
    documented: number
    partial: number
    undocumented: number
    implementation_verified: number
    implementation_declared: number
    implementation_missing: number
  }
  diagnostics: {
    structural_score: number
    status: 'healthy' | 'review' | 'needs_structure'
    findings: ProjectKnowledgeMapFinding[]
  }
  budget: {
    classification_model_tokens: 0
    markdown_bodies_read: 0
    metadata_only: true
  }
}

export interface ProjectKnowledgeMaps {
  capabilities: ProjectKnowledgeMap
  architecture: ProjectKnowledgeMap
  topics: ProjectKnowledgeMap
}

export const EMPTY_KNOWLEDGE_GRAPH: ProjectKnowledgeGraphConfig = { nodes: [], edges: [] }

export function sanitizeKnowledgeGraphConfig(value: unknown): ProjectKnowledgeGraphConfig {
  if (!value || typeof value !== 'object') return { nodes: [], edges: [] }
  const candidate = value as Partial<ProjectKnowledgeGraphConfig>
  const nodes = Array.isArray(candidate.nodes) ? candidate.nodes.slice(0, 256).flatMap((raw) => {
    if (!raw || typeof raw !== 'object') return []
    const node = raw as Partial<ProjectKnowledgeNodeConfig>
    const id = stableId(node.id, 80)
    const view = node.view === 'capabilities' || node.view === 'architecture' ? node.view : null
    const label = String(node.label ?? '').trim().slice(0, 60)
    if (!id || !view || !label) return []
    return [{
      id,
      view,
      kind: stableId(node.kind, 40) || (view === 'capabilities' ? 'capability' : 'component'),
      label,
      detail: String(node.detail ?? '').trim().slice(0, 240),
      parent_id: stableId(node.parent_id, 80),
      order: boundedNumber(node.order, 9_999),
      color: /^#[0-9a-f]{6}$/i.test(String(node.color ?? '')) ? String(node.color) : '#7f8fb3',
      entrypoint: normalizedPath(node.entrypoint),
      document_paths: stringArray(node.document_paths, 48).map(normalizedPath).filter(Boolean),
      implementation_refs: stringArray(node.implementation_refs, 48).map((item) => item.slice(0, 500)),
      tags: stringArray(node.tags, 24).map((item) => item.slice(0, 80)),
    }]
  }) : []
  const nodeIds = new Set(nodes.map((node) => node.id))
  const edges = Array.isArray(candidate.edges) ? candidate.edges.slice(0, 512).flatMap((raw) => {
    if (!raw || typeof raw !== 'object') return []
    const edge = raw as Partial<ProjectKnowledgeEdgeConfig>
    const id = stableId(edge.id, 100)
    const source = stableId(edge.source, 80)
    const target = stableId(edge.target, 80)
    if (!id || source === target || !nodeIds.has(source) || !nodeIds.has(target)) return []
    return [{
      id,
      source,
      target,
      relation: stableId(edge.relation, 40) || 'related_to',
      label: String(edge.label ?? '').trim().slice(0, 80),
    }]
  }) : []
  return {
    nodes: [...new Map(nodes.map((node) => [node.id, node])).values()],
    edges: [...new Map(edges.map((edge) => [edge.id, edge])).values()],
  }
}

function stableId(value: unknown, limit: number) {
  return String(value ?? '').trim().toLowerCase().replace(/[^a-z0-9._-]+/g, '').slice(0, limit)
}

function normalizedPath(value: unknown) {
  return String(value ?? '').trim().replace(/\\/g, '/')
}

function stringArray(value: unknown, limit: number) {
  return Array.isArray(value)
    ? value.filter((item): item is string => typeof item === 'string').slice(0, limit)
    : []
}

function boundedNumber(value: unknown, max: number) {
  const number = Number(value)
  return Number.isFinite(number) ? Math.min(max, Math.max(0, Math.floor(number))) : 0
}

File: server/src/project_document_knowledge_graph_model.rs
//! Shared manifest schema for product-capability and technical-architecture maps.

use anyhow::{bail, Result};
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

use crate::project_document_file_operation_model::normalize_document_path;

pub(crate) const MAX_KNOWLEDGE_GRAPH_NODES: usize = 256;
pub(crate) const MAX_KNOWLEDGE_GRAPH_EDGES: usize = 512;

#[derive(Debug, Clone, Serialize)]
pub(crate) struct ProjectKnowledgeMaps {
    pub capabilities: ProjectKnowledgeMap,
    pub architecture: ProjectKnowledgeMap,
    pub topics: ProjectKnowledgeMap,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct ProjectKnowledgeMap {
    pub version: u8,
    pub view: String,
    pub title: String,
    pub source: String,
    pub root_id: String,
    pub nodes: Vec<ProjectKnowledgeMapNode>,
    pub edges: Vec<ProjectKnowledgeMapEdge>,
    pub stats: KnowledgeMapStats,
    pub diagnostics: KnowledgeMapDiagnostics,
    pub budget: KnowledgeMapBudget,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct ProjectKnowledgeMapNode {
    pub id: String,
    pub view: String,
    pub kind: String,
    pub label: String,
    pub detail: String,
    pub color: String,
    pub parent_id: String,
    pub section_id: String,
    pub depth: usize,
    pub child_count: usize,
    pub order: i32,
    pub document_count: usize,
    pub document_paths: Vec<String>,
    pub entrypoint: String,
    pub entrypoint_source: String,
    pub coverage: Vec<KnowledgeMapCoverage>,
    pub missing_coverage: Vec<String>,
    pub documentation_status: String,
    pub implementation_refs: Vec<KnowledgeMapEvidence>,
    pub implementation_status: String,
    pub source: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct ProjectKnowledgeMapEdge {
    pub id: String,
    pub source: String,
    pub target: String,
    pub relation: String,
    pub label: String,
    pub configured: bool,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct KnowledgeMapCoverage {
    pub key: &'static str,
    pub label: &'static str,
    pub covered: bool,
    pub count: usize,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct KnowledgeMapEvidence {
    pub reference: String,
    pub verification: &'static str,
}

#[derive(Debug, Clone, Default, Serialize)]
pub(crate) struct KnowledgeMapStats {
    pub nodes: usize,
    pub configured_nodes: usize,
    pub documented: usize,
    pub partial: usize,
    pub undocumented: usize,
    pub implementation_verified: usize,
    pub implementation_declared: usize,
    pub implementation_missing: usize,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct KnowledgeMapDiagnostics {
    pub structural_score: u8,
    pub status: &'static str,
    pub findings: Vec<KnowledgeMapFinding>,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct KnowledgeMapFinding {
    pub code: &'static str,
    pub severity: &'static str,
    pub node_id: String,
    pub message: String,
    pub suggested_action: &'static str,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct KnowledgeMapBudget {
    pub classification_model_tokens: u8,
    pub markdown_bodies_read: u8,
    pub metadata_only: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct ProjectKnowledgeGraphConfig {
    #[serde(default)]
    pub nodes: Vec<ProjectKnowledgeNodeConfig>,
    #[serde(default)]
    pub edges: Vec<ProjectKnowledgeEdgeConfig>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct ProjectKnowledgeNodeConfig {
    pub id: String,
    pub view: String,
    #[serde(default)]
    pub kind: String,
    pub label: String,
    #[serde(default)]
    pub detail: String,
    #[serde(default)]
    pub parent_id: String,
    #[serde(default)]
    pub order: i32,
    #[serde(default = "default_color")]
    pub color: String,
    #[serde(default)]
    pub entrypoint: String,
    #[serde(default)]
    pub document_paths: Vec<String>,
    #[serde(default)]
    pub implementation_refs: Vec<String>,
    #[serde(default)]
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct ProjectKnowledgeEdgeConfig {
    pub id: String,
    pub source: String,
    pub target: String,
    #[serde(default)]
    pub relation: String,
    #[serde(default)]
    pub label: String,
}

pub(crate) fn normalize_graph_config(
    mut graph: ProjectKnowledgeGraphConfig,
) -> Result<ProjectKnowledgeGraphConfig> {
    if graph.nodes.len() > MAX_KNOWLEDGE_GRAPH_NODES
        || graph.edges.len() > MAX_KNOWLEDGE_GRAPH_EDGES
    {
        bail!("项目知识图谱超过安全上限");
    }
    graph.nodes = graph
        .nodes
        .into_iter()
        .map(normalize_node)
        .collect::<Result<Vec<_>>>()?;
    let node_ids = graph
        .nodes
        .iter()
        .map(|node| node.id.clone())
        .collect::<HashSet<_>>();
    let node_views = graph
        .nodes
        .iter()
        .map(|node| (node.id.clone(), node.view.clone()))
        .collect::<HashMap<_, _>>();
    if node_ids.len() != graph.nodes.len() {
        bail!("项目知识图谱节点 id 必须全局唯一");
    }
    validate_hierarchy(&graph.nodes, &node_ids)?;
    graph.edges = graph
        .edges
        .into_iter()
        .map(|edge| normalize_edge(edge, &node_ids, &node_views))
        .collect::<Result<Vec<_>>>()?;
    let edge_ids = graph
        .edges
        .iter()
        .map(|edge| &edge.id)
        .collect::<HashSet<_>>();
    if edge_ids.len() != graph.edges.len() {
        bail!("项目知识图谱关系 id 必须唯一");
    }
    Ok(graph)
}

pub(crate) fn merge_graph_config(
    current: ProjectKnowledgeGraphConfig,
    proposed: ProjectKnowledgeGraphConfig,
) -> Result<ProjectKnowledgeGraphConfig> {
    let mut nodes = current
        .nodes
        .into_iter()
        .map(|node| (node.id.clone(), node))
        .collect::<HashMap<_, _>>();
    nodes.extend(
        proposed
            .nodes
            .into_iter()
            .map(|node| (node.id.clone(), node)),
    );
    let mut edges = current
        .edges
        .into_iter()
        .map(|edge| (edge.id.clone(), edge))
        .collect::<HashMap<_, _>>();
    edges.extend(
        proposed
            .edges
            .into_iter()
            .map(|edge| (edge.id.clone(), edge)),
    );
    normalize_graph_config(ProjectKnowledgeGraphConfig {
        nodes: nodes.into_values().collect(),
        edges: edges.into_values().collect(),
    })
}

pub(crate) fn validate_graph_document_paths(
    graph: &ProjectKnowledgeGraphConfig,
    known_paths: &HashSet<String>,
) -> Result<()> {
    for node in &graph.nodes {
        for path in node
            .document_paths
            .iter()
            .chain((!node.entrypoint.is_empty()).then_some(&node.entrypoint))
        {
            if !known_paths.contains(&path.to_ascii_lowercase()) {
                bail!("知识图谱节点 {} 引用了目录中不存在的文档：{path}", node.id);
            }
        }
    }
    Ok(())
}

fn normalize_node(mut node: ProjectKnowledgeNodeConfig) -> Result<ProjectKnowledgeNodeConfig> {
    node.id = stable_id(&node.id, 80);
    node.view = match node.view.trim().to_ascii_lowercase().as_str() {
        "capabilities" | "architecture" => node.view.trim().to_ascii_lowercase(),
        _ => bail!("知识图谱节点 view 只支持 capabilities 或 architecture"),
    };
    node.kind = stable_id(&node.kind, 40);
    if node.kind.is_empty() {
        node.kind = if node.view == "capabilities" {
            "capability".to_string()
        } else {
            "component".to_string()
        };
    }
    node.label = truncate(node.label.trim(), 60);
    node.detail = truncate(node.detail.trim(), 240);
    node.parent_id = stable_id(&node.parent_id, 80);
    node.order = node.order.clamp(0, 9_999);
    if !is_hex_color(&node.color) {
        node.color = default_color();
    }
    node.entrypoint = optional_document_path(&node.entrypoint)?;
    node.document_paths = node
        .document_paths
        .into_iter()
        .take(48)
        .map(|path| normalize_document_path(&path))
        .collect::<Result<Vec<_>>>()?;
    node.document_paths.sort();
    node.document_paths.dedup();
    node.implementation_refs = bounded_strings(node.implementation_refs, 48, 500);
    node.tags = bounded_strings(node.tags, 24, 80);
    if node.id.is_empty() || node.label.is_empty() {
        bail!("知识图谱节点必须包含有效 id 和 label");
    }
    Ok(node)
}

fn normalize_edge(
    mut edge: ProjectKnowledgeEdgeConfig,
    node_ids: &HashSet<String>,
    node_views: &HashMap<String, String>,
) -> Result<ProjectKnowledgeEdgeConfig> {
    edge.id = stable_id(&edge.id, 100);
    edge.source = stable_id(&edge.source, 80);
    edge.target = stable_id(&edge.target, 80);
    edge.relation = stable_id(&edge.relation, 40);
    edge.label = truncate(edge.label.trim(), 80);
    if edge.relation.is_empty() {
        edge.relation = "related_to".to_string();
    }
    if edge.id.is_empty()
        || edge.source == edge.target
        || !node_ids.contains(&edge.source)
        || !node_ids.contains(&edge.target)
    {
        bail!("知识图谱关系引用了无效节点");
    }
    if node_views.get(&edge.source) != node_views.get(&edge.target) {
        bail!("知识图谱关系不能跨越产品功能图与技术架构图");
    }
    Ok(edge)
}

fn validate_hierarchy(
    nodes: &[ProjectKnowledgeNodeConfig],
    node_ids: &HashSet<String>,
) -> Result<()> {
    let parents = nodes
        .iter()
        .map(|node| (node.id.as_str(), node.parent_id.as_str()))
        .collect::<HashMap<_, _>>();
    let views = nodes
        .iter()
        .map(|node| (node.id.as_str(), node.view.as_str()))
        .collect::<HashMap<_, _>>();
    for node in nodes {
        if !node.parent_id.is_empty()
            && (!node_ids.contains(&node.parent_id)
                || views.get(node.parent_id.as_str()) != Some(&node.view.as_str()))
        {
            bail!("知识图谱节点 {} 引用了无效或跨视图父节点", node.id);
        }
        let mut cursor = node.id.as_str();
        let mut visited = HashSet::new();
        for depth in 0..=5 {
            if !visited.insert(cursor) {
                bail!("知识图谱层级存在循环：{}", node.id);
            }
            let parent = parents.get(cursor).copied().unwrap_or_default();
            if parent.is_empty() {
                break;
            }
            if depth == 5 {
                bail!("知识图谱层级最多支持 6 层：{}", node.id);
            }
            cursor = parent;
        }
    }
    Ok(())
}

fn optional_document_path(value: &str) -> Result<String> {
    if value.trim().is_empty() {
        Ok(String::new())
    } else {
        normalize_document_path(value)
    }
}

fn stable_id(value: &str, max: usize) -> String {
    value
        .trim()
        .to_ascii_lowercase()
        .chars()
        .filter(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '-' | '_' | '.'))
        .take(max)
        .collect()
}

fn bounded_strings(values: Vec<String>, count: usize, chars: usize) -> Vec<String> {
    let mut output = values
        .into_iter()
        .take(count)
        .map(|value| truncate(value.trim(), chars))
        .filter(|value| !value.is_empty())
        .collect::<Vec<_>>();
    output.sort();
    output.dedup();
    output
}

fn truncate(value: &str, limit: usize) -> String {
    value.chars().take(limit).collect()
}

fn is_hex_color(value: &str) -> bool {
    value.len() == 7
        && value.starts_with('#')
        && value[1..].chars().all(|ch| ch.is_ascii_hexdigit())
}

fn default_color() -> String {
    "#7f8fb3".to_string()
}
